API provides functionality for user to have different roles

Fraseer opened the pull request:
PT Link: https://www.pivotaltracker.com/story/show/179979042

## User Story
As a newspaper api,
in order to limit controller access
I would like to give users specific roles

## Changes Proposed
- Add pundit & pundit matcher
- Add a enum `roles` column to the `User` table and user model
- Editor allowed CRUD actions: Create, Read, Edit,& Destroy
- Journalist allowed CRUD actions: Create, Read, & Edit(Destroy own articles)
- Subscriber allowed CRUD actions: Read (index & show)
- Registered User allowed CRUD actions: Read(index)

- Update `create` and `show` actions in `articles_controller` to use Pundit authorize
- Update `application_controller` with `user_not_authorized` rescue  
- Update article model to have `has_and_belongs_to_many` association with authors 
- Update user model to have `has_and_belongs_to_many` association with articles 
- Create Article Policy and set up permissions for different roles
- Create `articles_authors_join_table`

## Screenshots
![image](https://user-images.githubusercontent.com/26599559/138544799-0cc10982-b6fd-45df-aacc-e2f0d2c6e03d.png)
![image](https://user-images.githubusercontent.com/26599559/138544813-960bbda2-f053-41b5-b292-bb8dc271638a.png)
![image](https://user-images.githubusercontent.com/26599559/138544836-290b8a91-efaa-4983-9ef2-9376b8afd79e.png)

Commit:
Creates seed files

Commit:
Adds pundit & matchers, generates app and created article policy. Seeds some articles and categories

tochman commented on spec/factories/user.rb:
```diff
@@ -0,0 +1,7 @@
+    password { '1234' }
+    role { :editor }
+  end
+end
```
```ruby
FactoryBot.define do
  factory :user do
    email { 'example@email.com' }
    password { '1234' }
    factory :editor do
       role { :editor }
     end
    factory :journalist do
       role { :journalist }
     end
  end
end
```

Commit:
Changes from SCRUM:
updated factory bot to have different roles
removed journalist from list of users
changed default to modern syntax

Commit:
tidy up spec file

Commit:
Updates policy to article and changes index and show to true

tochman committed:
updates user specs and article policies

tochman committed:
updates article policy and policy_spec

tochman committed:
configures habtm association articles authors/users

tochman committed:
cleans up the test for user

tochman committed:
configures application controller to use Pundit
refactors spec to reflect use of  policies

dorianbuck committed:
adds content for user roles

dorianbuck committed:
adds author_ids to params in create_article_spec

Commit:
Adds tests for user roles on create & show, updates articles controller

Commit:
merge from latest thomas, new tests for validation of presence in user models

maxarvid commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+    describe 'has many :articles' do
+      subject { create(:journalist) }
+      let(:user_2) { create(:journalist) }
```
It might be superfluous to have `user_2` for this particular test. We are not testing that an article can have many authors, but rather that the `subject`, in this case a journalist can be called with `.articles`

Fraseer replied:
Good point, probably want to test for multiple articles here instead

maxarvid commented on spec/policies/article_policies_spec.rb:
```diff
@@ -0,0 +1,70 @@
+      it { is_expected.to_not forbid_actions %i[update destroy] }
+    end
+
+    describe 'the user editor and not an author' do
```
Missing an `is` in this sentence

maxarvid commented on spec/policies/article_policies_spec.rb:
```diff
@@ -0,0 +1,70 @@
+      it { is_expected.to permit_actions %i[show index update destroy] }
+    end
+
+    describe 'the user editor and an author' do
```
same as above

maxarvid commented on spec/requests/api/create_article_spec.rb:
```diff
@@ -1,14 +1,21 @@
+  let(:credentials) { user.create_new_auth_token }
   let(:category) { create(:category) }
+  let(:user) { create(:journalist) }
+  let(:credentials) { user.create_new_auth_token }
```
I don't think these lines are necessary , they repeat line 3 and 5

Fraseer replied:
Removed

maxarvid commented on spec/requests/api/create_article_spec.rb:
```diff
@@ -1,14 +1,21 @@
 RSpec.describe 'POST /api/articles', type: :request do
   subject { response }
+  let(:user) { create(:journalist) }
+  let(:user_2) { create(:user) }
```
Looks to me as though `user_2` does not come into play below

Fraseer replied:
It does not

maxarvid commented on spec/user_roles/users_create_articles_spec.rb:
```diff
@@ -0,0 +1,56 @@
+RSpec.describe 'POST /api/articles', type: :request do
+  subject { response }
+  let(:category) { create(:category) }
+  let(:credentials) { user.create_new_auth_token }
```
This could lead to some false positives. Let's create a user so we can specify what credentials we are creating here

Fraseer replied:
So my thinking here was we create users within each of the describe blocks, each with different roles, and then have the same device auth credentials for each of them. This way every user is authorised in the device sense, and then its the role authorisations we are testing further on

Fraseer replied:
Im 99% sure it works as described

Fraseer replied:
Yeah so if you change the name of say the editor to something other than user, it then lacks the ability to make changes as it does not have the device auth credentials. Meaning that as each time we create a user in the describe blocks, called user, we are using the credentials created up the top

maxarvid replied:
If it works it works. In my mind the credentials pointed to a particular user in the `users` table. But if it's smart enough to check the roles independently then we are of course good.

maxarvid commented on spec/user_roles/users_show_article_spec.rb:
```diff
@@ -0,0 +1,51 @@
+RSpec.describe 'GET /api/articles', type: :request do
+  subject { response }
+  let(:category) { create(:category) }
+  let(:credentials) { user.create_new_auth_token }
```
Same as in the other users spec file. This could lead to false positives

Fraseer replied:
see above

maxarvid reviewed (changes requested):
Left a couple of things, mostly in the specs that I think need to be addressed.

Commit:
tidy up spec files for PR

maxarvid reviewed (approved):
LGTM

lukeperrin10 commented on spec/models/article_spec.rb:
```diff
@@ -3,9 +3,9 @@
-    it { is_expected.to have_db_column :category_id}
-    it { is_expected.to have_db_column :category_name}
-    it { is_expected.to have_db_column :published}
+    it { is_expected.to have_db_column :category_id }
+    it { is_expected.to have_db_column :category_name }
+    it { is_expected.to have_db_column :published }
```
Thought the formatter issues were fixed?

lukeperrin10 reviewed (approved):
Looks good. Just fix those for matters, couple of spaces added here and there which just makes for extra reading. Even an extra file in one example. 2 make sure you do some manual testing as well.spin up the rails server and point your client side to the local host. Do it for every feature. So there is no panic when a bug creeps in to production.
